pkg/network/ebpf/c/protocols/tls/tls.h

```diff
@@ -1,132 +1,459 @@
+#define COMPRESSION_METHODS_LENGTH   1  // Compression Methods length field is 1 byte
+#define EXTENSION_TYPE_LENGTH        2  // Extension Type field is 2 bytes
+#define EXTENSION_LENGTH_FIELD       2  // Extension Length field is 2 bytes
+
+// For single-byte fields (list lengths, etc.)
+#define SINGLE_BYTE_LENGTH           1
+
+// Minimum extension header length = Extension Type (2 bytes) + Extension Length (2 bytes) = 4 bytes
+#define MIN_EXTENSION_HEADER_LENGTH (EXTENSION_TYPE_LENGTH + EXTENSION_LENGTH_FIELD)
+
+// Maximum number of supported versions we unroll for
+#define MAX_SUPPORTED_VERSIONS 4
+
+// TLS record layer header structure (RFC 5246)
 typedef struct {
     __u8 content_type;
     __u16 version;
     __u16 length;
 } __attribute__((packed)) tls_record_header_t;
 
-typedef struct {
-    __u8 handshake_type;
-    __u8 length[3];
+// Checks if the TLS version is valid
+static __always_inline bool is_valid_tls_version(__u16 version) {
+    switch (version) {
+        case SSL_VERSION20:
+        case SSL_VERSION30:
+        case TLS_VERSION10:
+        case TLS_VERSION11:
+        case TLS_VERSION12:
+        case TLS_VERSION13:
+            return true;
+        default:
+            return false;
+    }
+}
+
+// set_tls_offered_version sets the bit corresponding to the offered version in the offered_versions field of tls_info
+static __always_inline void set_tls_offered_version(tls_info_t *tls_info, __u16 version) {
+    switch (version) {
+        case TLS_VERSION10:
+            tls_info->offered_versions |= TLS_VERSION10_BIT;
+            break;
+        case TLS_VERSION11:
+            tls_info->offered_versions |= TLS_VERSION11_BIT;
+            break;
+        case TLS_VERSION12:
+            tls_info->offered_versions |= TLS_VERSION12_BIT;
+            break;
+        case TLS_VERSION13:
+            tls_info->offered_versions |= TLS_VERSION13_BIT;
+            break;
+        default:
+            break;
+    }
+}
+
+// read_tls_record_header reads the TLS record header from the packet
+static __always_inline bool read_tls_record_header(struct __sk_buff *skb, __u64 header_offset, __u32 data_end, tls_record_header_t *tls_hdr) {
+    // Ensure there's enough space for TLS record header
+    if (header_offset + sizeof(tls_record_header_t) > data_end) {
+        return false;
+    }
+
+    // Read TLS record header
+    if (bpf_skb_load_bytes(skb, header_offset, tls_hdr, sizeof(tls_record_header_t)) < 0) {
+        return false;
+    }
+
+    // Convert fields to host byte order
+    tls_hdr->version = bpf_ntohs(tls_hdr->version);
+    tls_hdr->length = bpf_ntohs(tls_hdr->length);
+
+    // Validate version and length
+    if (!is_valid_tls_version(tls_hdr->version)) {
+        return false;
+    }
+    if (tls_hdr->length > TLS_MAX_PAYLOAD_LENGTH) {
+        return false;
+    }
+
+    // Ensure we don't read beyond the packet
+    return header_offset + sizeof(tls_record_header_t) + tls_hdr->length <= data_end;
+}
```

Comment:
The original implementation we had was simpler, and this version does not provide any improvement
I do think we should revert that

Reply:
discussed [here](https://github.com/DataDog/datadog-agent/pull/31464#discussion_r1890494237)